packages/blockchain/test/4444Optimistic.spec.ts

```diff
@@ -0,0 +1,308 @@
+    let dbBlock, headBlock
+    const blockchain = await createBlockchain({ genesisBlock, common, validateBlocks: false })
+
+    for (const block of [block1, block2, block3PoS, block4, block5, block6, block7, block8]) {
+      await blockchain.putBlock(block, { canonical: false }).catch((_e) => {
+        undefined
+      })
+    }
+
+    for (const block of [block1, block2, block3PoS, block4, block5, block6, block7, block8]) {
+      dbBlock = await blockchain.getBlock(block.hash()).catch((_e) => undefined)
+      assert.equal(dbBlock !== undefined, true, 'block should be inserted')
+
+      const td = await blockchain.getTotalDifficulty(block.hash()).catch((_e) => undefined)
+      assert.equal(td !== undefined, true, 'block should be marked complete')
+
+      // however the block by number should still not fetch it
+      dbBlock = await blockchain.getBlock(block.header.number).catch((_e) => undefined)
+      assert.equal(dbBlock === undefined, true, 'block number index should not exit')
+    }
+
+    // head should not move
+    headBlock = await blockchain.getCanonicalHeadBlock()
+    assert.equal(
+      headBlock.header.number,
+      genesisBlock.header.number,
+      'head should still be at genesis',
+    )
+
+    // however normal insertion should respect canonicality updates
+    for (const block of [block1, block2, block3PoS, block4, block5, block6, block7, block8]) {
+      await blockchain.putBlock(block).catch((_e) => {
+        undefined
+      })
+    }
+    headBlock = await blockchain.getCanonicalHeadBlock()
+    assert.equal(headBlock.header.number, block8.header.number, 'head should still be at genesis')
+
+    for (const block of [block1, block2, block3PoS, block4, block5, block6, block7, block8]) {
+      // however the block by number should still not fetch it
+      dbBlock = await blockchain.getBlock(block.header.number).catch((_e) => undefined)
+      assert.equal(dbBlock !== undefined, true, 'block number index should not exit')
+    }
+
+    // since block61 doesn't has the parent block51 inserted yet, they will still be inserted as non
+    // canonical
+    for (const block of [block61, block71]) {
+      await blockchain.putBlock(block).catch((_e) => {
+        undefined
+      })
+    }
+    // verify
+    headBlock = await blockchain.getCanonicalHeadBlock()
+    assert.equal(headBlock.header.number, block8.header.number, 'head should still be at genesis')
+
+    for (const block of [block61, block71]) {
+      // however the block by number should still not fetch it
+      dbBlock = await blockchain.getBlock(block.header.number).catch((_e) => undefined)
+      assert.equal(
+        equalsBytes(dbBlock.hash(), block.hash()),
+        false,
+        'block number index should not exit',
+      )
+
+      const td = await blockchain.getTotalDifficulty(block.hash()).catch((_e) => undefined)
+      assert.equal(td, undefined, 'block should be not be marked complete')
```

Comment:
a complete block is the one which is confirmed to be linked all the way to the genesis via chain of parent blocks.

in our blockchain db, this is marked via block having the totalDifficulty written onto the chain. a boolean flag could also be used, but total difficulty serves well for both PoW and PoS blocks since PoW total difficulty is actually need to make head and canonicality decisions, while for PoS it just acts an an indicator.

while putting a block, if block's parent is in the blockchain and the parent also has totalDifficulty, then the new block is considered complete and it can move head

Reply:
We should really really define (write down!) very clearly what we mean with both `canonical` and `complete` here. Again, maybe we are also "not there" with these names yet (e.g. `complete` for a block with this kind of definition is very misleading, one is just thinking in the realm of that the block has all txs, fields, .... We need to urgently find some better and more precise terminology here).